src/librustc_mir/interpret/cast.rs

```diff
@@ -47,7 +46,19 @@ impl<'a, 'mir, 'tcx, M: Machine<'a, 'mir, 'tcx>> EvalContext<'a, 'mir, 'tcx, M>
 
             Misc => {
                 let src = self.read_value(src)?;
-                if self.type_is_fat_ptr(src_layout.ty) {
+
+                if src.layout.ty.is_region_ptr() && dest.layout.ty.is_unsafe_ptr() {
```

Comment:
slightly unhappy about code that isn't needed during const eval but still will be compiled into const eval. Not sure what to do about that though

Reply:
I cannot make it depend on `enforce_validity()` as that whitelists some functions in miri and I think my model might go crazy if references are only sometimes tracked.

I could add another `const REF_TRACKING_HOOKS: bool` though.

Reply:
Good idea, not a blocker for this PR though